xdsl/transforms/function_transformations.py

```diff
@@ -36,6 +40,70 @@ def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
             rewriter.has_done_action = True
 
 
+TIMER_START = "timer_start"
+TIMER_END = "timer_end"
+
+
+@dataclass
+class AddBenchTimersPattern(RewritePattern):
+    start_func_t: func.FunctionType
+    end_func_t: func.FunctionType
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
+        if (
+            not (top_level := op.parent_op())
+            or not isinstance(top_level, builtin.ModuleOp)
+            or top_level.parent
+        ):
+            return
+
+        ptr = op.body.block.insert_arg(llvm.LLVMPointerType.opaque(), len(op.args))
+        start_call = func.Call(TIMER_START, [], tuple(self.start_func_t.outputs))
+        end_call = func.Call(TIMER_END, start_call.res, tuple(self.end_func_t.outputs))
+        store_time = llvm.StoreOp(end_call.res[0], ptr)
+
+        ptr.name_hint = "timers"
+        start_call.res[0].name_hint = "timestamp"
+        end_call.res[0].name_hint = "timediff"
+
+        assert op.body.block.last_op
+        rewriter.insert_op(start_call, InsertPoint.at_start(op.body.block))
+        rewriter.insert_op(
+            [end_call, store_time], InsertPoint.before(op.body.block.last_op)
+        )
+        op.update_function_type()
+
+
+class TestAddBenchTimersToTopLevelFunctions(ModulePass):
+    """
+    Adds timers to top-level functions, by adding `timer_start() -> f64` and `timer_end(f64) -> f64`
+    to the start and end of each module-level function. The time is stored in an `llvm.ptr` passed in
+    as a function arg.
+    """
+
+    name = "test-add-timers-to-top-level-funcs"
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        all_funcs = [f for f in op.body.block.ops if isinstance(f, func.FuncOp)]
+        func_names = [f.sym_name.data for f in all_funcs]
+        if TIMER_START in func_names or TIMER_END in func_names:
+            return
+
+        start_func_t = func.FunctionType.from_lists([], [builtin.Float64Type()])
+        end_func_t = func.FunctionType.from_lists(
+            [builtin.Float64Type()], [builtin.Float64Type()]
+        )
+        start_func = func.FuncOp(TIMER_START, start_func_t, Region([]))
+        end_func = func.FuncOp(TIMER_END, end_func_t, Region([]))
+
+        PatternRewriteWalker(
+            AddBenchTimersPattern(start_func_t, end_func_t), apply_recursively=False
+        ).rewrite_module(op)
+
+        op.body.block.add_ops((start_func, end_func))
```

Comment:
I would recommend using SymbolTable APIs to look for functions and insert them:

xdsl/backend/riscv/lowering/convert_memref_to_riscv.py

Reply:
Sweet, I like this